0000-scheduled-circuit.md

```diff
@@ -0,0 +1,266 @@
+  - O(W): update end time dictionary
+- Absolutely-timed:
+  - O(W): fill last idle time-slot with delays
+  - O(N): extend lists by qubit
+  - O(N): update start times
+
+2) replace(refs: Positions, targets: LengthedInstructions)*
+    - fail if refs and targets has different duration
+- Relatively-timed:
+  - O(N): check validity of arguments
+  - O(ListReplace(D)*W): replace items on lists by qubit
+- Absolutely-timed:
+  - O(N): check validity of arguments
+  - O(N): update start times
+  - O(ListReplace(D)\*W): replace items on lists by qubit
+
+3) insert(ref: Position, targets: LengthedInstructions)*
+    - fail if no enough idle time before the reference instruction
+- Relatively-timed:
+  - O(W): check validity of arguments
+  - O(ListInsert(D)*W): insert items on lists by qubit
+- Absolutely-timed:
+  - O(W): check validity of arguments
+  - O(N): update start times
+  - O(ListInsert(D)*W): insert items on lists by qubit
+
+4) insert_at_time(time: Time, targets: LengthedInstructions)
+    - fail if no enough idle time between the time and start time of next instruction
+- Relatively-timed:
+  - O(D*W): overlap check
+  - O(ListReplace(D)*W): replace items on lists by qubit
+- Absolutely-timed:
+  - O(log(D)*W): overlap check
+  - O(N): update start times
+  - O(ListReplace(D)*W): replace items on lists by qubit
+
+In addition to these, we need kinds of “find()” family for users to specify references in insert/replace etc.
+
+\*) I assume the interface that references are given as Postions.
+There is a clear drawback in referring by positions:
+replacement of cx gate, which would be specified by two Positions and may confuse users.
+(*To achieve single position for cx gate, it would be better to use DAG data structure like DAGCircuit…*).
+If references should be given as Instructions, we need to accept overhead to find positions.
+Such overhead would be O(D\*W) for Relatively-timed and O(log(D)*W) for Absolutely-timed unless we add another data structure for the purpose.
+
+
+### Dictionary vs DAG data structure
+How to store *ordering* among instructions is much important point for selecting data structure.
+This is a problem independent of Relatively-timed or Absolutely-timed, which addresses the problem if we should keep absolute-timing in data or not.
+
+We can think of two data structure for storing the orders among instructions:
+- Dictionary (Dict[Qubit, List[LengthedInstruction]] discussed above)
+- DAG (as like used in DAGCircuit)
+
+Note: *For any node of DAGCircuit, its in-(or out-)edges must have different qubit label.
+It enables DAGCircuit to store the ordering of instruction by qubits.*
+As for general DependencyDAG, which represents dependencies among instructions,
+its node can have multiple in-(or out-)-edges with the same qubit label.
+So DependencyDAG is not appropriate to store the ordering of instruction by qubits.
+
+In Dictionary, any multi-qubit instruction is stored at multiple positions on multiple lists.
```

Comment:
For this reason I do not believe a dictionary is very viable, maintaining these qubit dictionaries I think is equivalent to maintaining a list of integers for each qubit, where the integer specifies the index of the instruction in the `instructions` list. These must be updated on the append/removal of an instruction making these operations O(j*k) where  is the average number of qubits and k the average number of instructions per qubit. This could be slightly mitigated if instructions could be hashed, which may not be true in general.

Consequently I think the two clear choices for an IR structure (as in most IRs) are the linear IR (a list of instructions) or a DAG. I think the base IR makes sense to be linear as it is much simpler and if necessary we can support the DAG version via a conversion.